### PR TITLE
Change default value of ONE_DIR

### PR DESCRIPTION
**Pull request by casperklein**
# Description

This changes the default value of `ONE_DIR` from `0` to `1`.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Improvement (non-breaking change that does improve existing functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (README.md or the documentation under `docs/`)
- [ ] If necessary I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -23,7 +23,7 @@ LABEL org.opencontainers.image.source="https://github.com/docker-mailserver/dock
 
 ENV ENABLE_POSTGREY=0
 ENV FETCHMAIL_POLL=300
-ENV ONE_DIR=0
+ENV ONE_DIR=1
```
Comment (casperklein) [inline Dockerfile:26]: Does anyone knows, whats the purpose of having this ENV here?

Comment (georglauterbach) [reply Dockerfile:26]: I think this variable is read in some scripts / needed somewhere even before it is read in `start-mailserver.sh`.

**Review by georglauterbach — APPROVED** [review]

File: Dockerfile (hunk 1 above)
Comment (casperklein) [reply Dockerfile:26]: > somewhere even before it is read in start-mailserver.sh

Startup: `dumb-init --> supervisor --> start-mailserver.sh`

I couldn't find / think of a place that comes before `start-mailserver.sh`. The same goes as well for the other ENVs.
Maybe in a separate PR we can give it a try (removing the ENVs) and see if tests pass 😃

**Comment by casperklein** [comment]
The build-and-test check fails due a temporary clamav issue. This is not related to changes in PR.

**Review by polarathene — APPROVED** [review]
Just copying this over from maintainer discussion to provide additional context behind the decision for the change, in case anyone stumbles upon this PR and wonders why it was done:

> The only downside I can think of is the additional disk space used.
> 
> But when using fail2ban for example, this preserves the bans during container restarts. See also this [comment](https://github.com/docker-mailserver/docker-mailserver/pull/2144#discussion_r693563902)
>
> If I remember correctly, ClamAV will take some good amount of space. My mail-state is currently arround ~24MB (without ClamAV)

Most notably the linked comment mentions `ONE_DIR=1` places all state into `/var/mail-state` so that state like Fail2Ban can be preserved across restarts. That's usually desirable for this service, so makes a good default.

**Comment by polarathene** [comment]
> fails due a temporary clamav issue.

Are you sure it's temporary? The error message seems to make it rather clear that it'll continue to be an issue. I don't think it matters how much we run the CI test, but that it'd be from Github IPs that would be getting blocked?:

```
#90 0.411 Mon Aug 23 19:43:50 2021 ->     Ensure you are the most updated version by visiting https://www.clamav.net/downloads
#90 0.411 Mon Aug 23 19:43:50 2021 ->  2. Your network is explicitly denied by the FreshClam CDN.
#90 0.411 Mon Aug 23 19:43:50 2021 ->     In order to rectify this please check that you are:
#90 0.411 Mon Aug 23 19:43:50 2021 ->    a. Running an up-to-date version of FreshClam
#90 0.411 Mon Aug 23 19:43:50 2021 ->    b. Running FreshClam no more than once an hour
#90 0.411 Mon Aug 23 19:43:50 2021 ->    c. If you have checked (a) and (b), please open a ticket at
#90 0.411 Mon Aug 23 19:43:50 2021 ->       https://bugzilla.clamav.net under the 'Mirrors' component
#90 0.411 Mon Aug 23 19:43:50 2021 ->       and we will investigate why your network is blocked.
#90 0.411 Mon Aug 23 19:43:50 2021 -> !Database update process failed: Forbidden; Blocked by CDN
#90 0.411 Mon Aug 23 19:43:50 2021 -> !Update failed.
```

[Related upstream issue](https://github.com/clamwin/clamav/issues/4):

> ..frequently download the whole database set, or are deploying containers or VMs that don't contain a baseline set of databases and thus when started download the whole database sets.
>
> This has become increasingly costly for the ClamAV project. So the ClamAV project is making a best effort to require users to use FreshClam to update existing databases and to not download the entire database set unless absolutely necessary.
>
> To that end, programs like wget have been blocked entirely (error code 403) and the whole database files like daily.cvd, main.cvd, and bytecode.cvd are being rate limited (error code 429).
>
> The update patch files (the *.cdiff files) are not rate limited, so good netizens can easily update their existing ClamAV installs by running FreshClam.

If it continues to be a problem we might need a workaround (perhaps a separate base image as a cache). I'll restart the test suite.

**Comment by casperklein** [comment]
> Are you sure it's temporary?
> but that it'd be from Github IPs that would be getting blocked?

I also encounter this, when building DMS locally. So the error is not related to Github IPs.

**Comment by polarathene** [comment]
> I'll restart the test suite.

Seems we have a test failure, could it be related to this change? The test doesn't seem to have used `ONE_DIR=1` with the container:

```
not ok 299 checking quota: warn message received when quota exceeded
# (from function `assert_output' in file test/test_helper/bats-assert/src/assert.bash, line 239,
#  in test file test/tests.bats, line 902)
#   `assert_output "2"' failed
# 
# -- output differs --
# expected : 2
# actual   : 3
# --
```

https://github.com/docker-mailserver/docker-mailserver/blob/13e8c798a6aa2c08598792c99f0d40748cd1c74d/test/tests.bats#L899-L902

https://github.com/docker-mailserver/docker-mailserver/blob/13e8c798a6aa2c08598792c99f0d40748cd1c74d/test/tests.bats#L16-L37

**Comment by georglauterbach** [comment]
I re-ran CI yesterday and the error is still present. So yeah, the error does not seem to be temporary.

**Comment by casperklein** [comment]
> The test doesn't seem to have used ONE_DIR=1 with the container

If ONE_DIR is not defined, it now defaults to 1. Before it was 0. I'll add ONE_DIR=0 and see if that makes a change.

**Review by polarathene — DISMISSED** [review]
Tests have passed, but this is also an indication that the test is not covering `ONE_DIR=1` which we want to make the default, it may be worth looking into fixing that?

**Comment by casperklein** [comment]
I am not that familiar with our tests. But in theory, there should be no difference in how DMS services works, if ONE_DIR is set or not.

**Comment by georglauterbach** [comment]
> You can see the failed test here: https://github.com/docker-mailserver/docker-mailserver
> 
> 
> 
> ```
> 
> not ok 299 checking quota: warn message received when quota exceeded
> 
> # (from function `assert_output' in file test/test_helper/bats-assert/src/assert.bash, line 239,
> 
> ...

The quota test has always been an issue, and we can "safely" restart the test suite if this test fails. The tests seem to pass ATM.

**Comment by casperklein** [comment]
>  The tests seem to pass ATM.

Because of https://github.com/docker-mailserver/docker-mailserver/pull/2148/commits/8cc892cf0178c15062bd1dc26bd6ca7eddc3ad18

I'll revert that to see if the test still fails.

**Comment by georglauterbach** [comment]
> >  The tests seem to pass ATM.
> 
> 
> 
> Because of https://github.com/docker-mailserver/docker-mailserver/pull/2148/commits/8cc892cf0178c15062bd1dc26bd6ca7eddc3ad18
> 
> 
> 
> I'll revert that to see if the test still fails.

Ah, I see. But test still pass :)

**Review by georglauterbach — APPROVED** [review]

**Comment by casperklein** [comment]
As our tests pass, with `ONE_DIR'` set to `0` or `1`, I think this can be merged?

**Comment by georglauterbach** [comment]
> As our tests pass, with `ONE_DIR'` set to `0` or `1`, I think this can be merged?

Agreed.

**Comment by NorseGaud** [comment]
@georglauterbach @casperklein this is a pretty large change for existing users right? Maybe we should wait until a minor version bump?

**Comment by georglauterbach** [comment]
This is built for `:edge` anyway. Users using `:edge` _should_ be aware of certain changes. But if wouldn't argue about releasing `v10.1.3` if you'd like to do this afterwards.

**Review by polarathene — APPROVED** [review]

**Comment by casperklein** [comment]
As @georglauterbach wrote, merging only affects the :edge build. Users with _versioned_ or _latest_ tags are not affected. Only, once 10.2.0 is released.

**Comment by georglauterbach** [comment]
@casperklein I think you can merge this. It would be interesting to see what happens when the `ENV`s are removed from the Dockerfile. A separate PR could be opened for that.

**Pull request merged**